### PR TITLE
Upload scripts

### PR DESCRIPTION
**Pull request by squidsoup**
## Done
Allow uploading of testing and commissioning scripts.

## QA
1. Visit either testing or commissioning scripts.
2. Click "Upload script"
3. Ensure, Cancel takes you back to the appropriate script list view.
4. Click "Upload script" again.
5. Drag and drop a valid script (text file < 2MB)
6. Click upload
7. You should be redirected to the scripts list view, see your script in the list, and see a success message. Ensure the contents of the script are displaying when clicking the script name.
8. Click "Upload script" again.
9. Try out various cases for invalid files (> 2MB, wrong mimetype)
10. Try dragging dropping more than one file etc
You should see appropriate errors explaining whats supported.

## Also Fixes
https://github.com/canonical-web-and-design/maas-ui/issues/184

### CODE REVIEW

File: ui/src/app/base/components/MainTable/MainTable.js [hunk 1]
```diff
@@ -189,7 +189,7 @@ const MainTable = ({
             sortable
           )}
       </Table>
-      {paginate && rows && rows.length && (
+      {paginate && rows && rows.length > 0 && (
```
Comment (squidsoup) [inline ui/src/app/base/components/MainTable/MainTable.js]: Driveby fix for https://github.com/canonical-web-and-design/maas-ui/issues/184

File: ui/src/app/base/sagas/http.js [hunk 2]
```diff
@@ -47,9 +59,29 @@ export function* fetchScriptsSaga() {
...
+    });
+  }
+}
+
+export function* uploadScriptSaga(action) {
+  const csrftoken = yield call(getCookie, "csrftoken");
+  const script = action.payload;
+  let response;
+  try {
+    yield put({ type: `UPLOAD_SCRIPT_START` });
```
Comment (huwshimi) [inline ui/src/app/base/sagas/http.js]: There's a heap of places in these files that use the `` form of strings, I'm surprised prettier doesn't change that.

File: ui/src/app/settings/views/Scripts/ScriptsUpload.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+import { messages } from "app/base/actions";
+import Card from "app/base/components/Card";
+import Form from "app/base/components/Form";
+import FormCardButtons from "app/base/components/FormCardButtons";
+import Row from "app/base/components/Row";
+import actions from "app/settings/actions";
+import selectors from "app/settings/selectors";
+
+const baseStyle = {
```
Comment (huwshimi) [inline ui/src/app/settings/views/Scripts/ScriptsUpload.js]: I know we talked about this, but seeing how many styles you're adding here I think this would be best as a few classes in a CSS file. One of the benefits will be that the colours can use vanilla variables.

File: ui/src/app/settings/views/Scripts/ScriptsUpload.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  borderRadius: 2,
+  borderColor: "#eeeeee",
+  borderStyle: "dashed",
+  backgroundColor: "#fafafa",
+  color: "#bdbdbd",
+  outline: "none",
+  transition: "border .24s ease-in-out"
+};
+
+const activeStyle = {
```
Comment (huwshimi) [inline ui/src/app/settings/views/Scripts/ScriptsUpload.js]: These would become something like `scripts-upload--active` and use the classnames dep to apply the right thing.

File: ui/src/app/settings/views/Scripts/ScriptsUpload.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+          messages.add(`Invalid filetype, please try again.`, "negative")
+        );
+        return;
+      }
+
+      const acceptedFile = acceptedFiles[0];
+      const scriptName = pathParse(acceptedFile.path).name;
+      const reader = new FileReader();
+      reader.onabort = () => console.error("file reading was aborted");
+      reader.onerror = () => console.error("file reading has failed");
```
Comment (huwshimi) [inline ui/src/app/settings/views/Scripts/ScriptsUpload.js]: Should these show an "Uploading failed" or something message to the user?

File: ui/src/app/settings/views/Scripts/ScriptsUpload.js [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    }
+  }, [dispatch, saved, savedScript]);
+
+  if (saved) {
+    // The script was successfully uploaded so redirect to the scripts list.
+    return <Redirect to={`/scripts/${type}`} />;
+  }
+
+  return (
+    <Card>
```
Comment (huwshimi) [inline ui/src/app/settings/views/Scripts/ScriptsUpload.js]: It might be nice to update the `FormCard` to handle the case where the heading is on its own row and use that here.

File: ui/src/app/settings/views/Scripts/ScriptsUpload.js [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        >
+          {acceptedFiles.length > 0 && (
+            <small>
+              {`${acceptedFiles[0].path} (${acceptedFiles[0].size} bytes) ready for upload.`}
+            </small>
+          )}
+
+          <FormCardButtons
+            actionDisabled={acceptedFiles.length === 0}
+            actionLabel={`Upload script`}
```
Comment (huwshimi) [inline ui/src/app/settings/views/Scripts/ScriptsUpload.js]: This can just be `actionLabel="Upload script"`

**Comment by huwshimi** [comment]
One QA issue: when uploading a script with metadata and a different name to the filename it complains about: "Error uploading test: May not override values defined in embedded YAML."

e.g. if a script was called `test.sh` and contained the following:
```shell
#!/usr/bin/env python3

# --- Start MAAS 1.0 script metadata ---
# name: nottest
# parallel: instance
# parameters:
#   storage: {type: storage}
# --- End MAAS 1.0 script metadata ---

import argparse

parser = argparse.ArgumentParser(description='')
parser.add_argument(
    '--storage', dest='storage', required=True,
    help='path to storage device you want to test. e.g. /dev/sda')
args = parser.parse_args()

print("Testing: %s" % args.storage)
```

This isn't an issue with the old UI, so I wonder if it might be extracting the name from the metadata.

**Comment by squidsoup** [comment]
> One QA issue: when uploading a script with metadata and a different name to the filename it complains about: "Error uploading test: May not override values defined in embedded YAML."
> 
> e.g. if a script was called `test.sh` and contained the following:
> 
> ```shell
> #!/usr/bin/env python3
> 
> # --- Start MAAS 1.0 script metadata ---
> # name: nottest
> # parallel: instance
> # parameters:
> #   storage: {type: storage}
> # --- End MAAS 1.0 script metadata ---
> 
> import argparse
> 
> parser = argparse.ArgumentParser(description='')
> parser.add_argument(
>     '--storage', dest='storage', required=True,
>     help='path to storage device you want to test. e.g. /dev/sda')
> args = parser.parse_args()
> 
> print("Testing: %s" % args.storage)
> ```
> 
> This isn't an issue with the old UI, so I wonder if it might be extracting the name from the metadata.

https://github.com/canonical-web-and-design/maas-ui/issues/209
